README.md
# XtsTinyBasicPlus

a fork of -TinyBasic for Arduino- by Xtase (fgalliat) @Nov 2016

It actually uses ~43.6KB of flash & 7K of RAM (statically allocated)<br/>
so It **WILL NOT run on Arduino INO boards** ...

(It's an early stage work, so it's dirty & not well documented)

* added support for :
   - ssd1306 LCD I2C Screen + 3 user btns (up/down/ok) **[usable inside the BASIC]**
   - SPI uSD Card reader
   - ESP8266 Wifi module
   - 2nd UART (two serial ports control)
   - Teensy ++2 module
   
   - Canon X-07 Serial protocol

![Dev Board](/images/devBoard.jpg)

* 3 user buttons & LCD control routines ( BTN() + LCCLS, LCPRINT &lt;line&gt;,&lt;expr&gt; ),<br/>
  You are able to create simple menus !!

* Newly created mainboard & started to build enclosure...
![Dev Board](/images/enclosure_plus_photo.jpg)

* It provides a regular (even if limited) BASIC shell, and can act as an HTTP Server to launch a 'web.bas' file<br/>
  that outputs directly in the browser.

```
10 FOR I =1 TO 5
20 PRINT "web ";: PRINT I
30 NEXT I
40 FILES
50 PRINT "Free Mem:";: PRINT FREE()
```

* It also have 2 serial ports (MASTER / SLAVE) that can be hot-swapped with 'CINV' command.<br/>
  note that 'CINIT' <baud_rate> does ONLY set the "Software" Serial port & that the "Hardware" Serial port (USB),<br/>
  is always @ 115200 bauds (you can change it in the code).<br/>

![Web + 2 serial ports control](/images/Web_2Serials.jpg)

* Since 11/28/2016 : Started an experimental way to produce 4 String variables

* Since 12/03/2016 : **Support now (w/ a few bugs anyway) Canon X-07 PRGs writing to/reading from uSDCard**<br/>
  See basic_source/loader.bas & X07_source/LOADSD.TXT & SV2SD.TXT
![Canon X-07 & the 'black box'](/images/X07_Teensy.jpg)

* previous authors of TinyBasic :
   - Mike Field <hamster@snap.net.nz>
   - Scott Lawrence <yorgle@gmail.com>
   - Brian O'Dell <megamemnon@megamemnon.com>
   
   - https://github.com/BleuLlama/TinyBasicPlus


docs/CHANGELOG.TXT
// va.15: 2016-12-03
//      NOW ABLE to load a PRG from SD to X-07
//           with [XTBP] : LOADER.BAS
//                [X-07] : RUN "LOADSD"
//
//      ADDED WRITETXT <filename> => prompts for text (w/ no ECHO)
//                                => a blank line closes the file
//                                => by default : this cmd appends to existing file
//                                => U will have to DELETE <filaname> before if want a clean file
//
//      NOW ABLE to save a PRG from X-07 to SD w/ some bugs ;(
//                [X-07] : RUN "SV2SD"
//
// va.15: 2016-12-02
//
//      CONNECTED & TUNED a bit the code in order to work w/ Canon X-07 SerialPort
//        it works well @ 8000bps ! Yeah :) (but need a between-char pause ;( )
//        have to find an equiv to |save "toto.bas"| (from the X-07 to uSD) .. in progress ..
//
//      ADDED parse of $1-$4 in filename expressions
//
// va.15: 2016-12-01
//      ADDED OUT x -> that outputs chr(x) to MASTER Uart (as COUT x)
//      FIXED CPRINT <expr> now supports '$1' to '$4'
//      ADDED INP() -> that waits a char FROM MASTER Uart (as a=CINP() )
//
//      -> TODO : CINPUTSTR 1-4
//      -> TODO : VARCHAR assignment (@ this time : only by INPUTSTR)
//      -> TODO : VARCHAR Ops : @least concat by '+'
//
//      -> cos, sin, tan functions w/ predefined values tables in degrees (use PROGMEM)
//      -> LCD plot pixel ???
//
// va.15: 2016-11-30
//      ADDED support for X07 UART mode switch (DO not manipulate when board is already powered on)
//      
// va.15: 2016-11-29
//      FIXED wifi enabled/disabled status detection
//  --- Currently Experimental ---
//      ALLOCATED 4 mem segments of 50bytes each for new string variables
//      ADDED 4 String variables '$1' to '$4'
//      MODIFIED STRLEN(1-4) func -> returns the len the '$1' to '$4' String variables -> (if no params : $1)
//      MODIFIED INPUTSTR 1-4 : input an str that goes in '$1' to '$4'  -> (if no params : $1)
//
//      $1-$4 parsed in get_quoted_string()
//                      print_quoted_string()
//                      lcd_print_quoted_string()
// 
//  --- Currently Experimental ---
//
//
// va.15: 2016-11-28
//      ADDED : WRITE "<file>","<quoted_string\nable>" : appends "quoted_string" to "file"
//                N.B.: can DELETE <file> before
//                N.B.: Oups : TinyBasic doesn't support char variables ....
//
//              INPUTSTR : input an str that goes in a unik str variable (50 bytes long max)
//                to recall => use '$'
//                works in get_quoted_string() .... -> LATER -> print_quoted_string() // lcd_print_quoted_string()
//
//              STRLEN() func -> returns the len (as cannot yet test the '$' 's value) (no arg @ this time)
//
//
// va.15: 2016-11-26
//      ADDED : BTN() func that returns 3-user-btns state if supported (else -1)
//                    1 for UP / 2 for DOWN / 4 for OK
//
//              LCCLS                 erase the LCD if defined
//              LCPRINT <line>,<expr> write on LCD @line (1 to 4) if defined
//
//              so now, You're able to create simple menus !!
//
// va.15: 2016-11-25 Xtase (fgalliat) fork
//      ADDED : free() function + way to have no-arg function
//      ADDED : WRESET to reset the ESP8266
//      ADDED : DELETE "<file>" to remove filr from sdcard
//      ADDED : CAT "<file>" to list a file from sdcard w/o interpret it
//
//      CHANGED BEHAVIOUR of FILES -> no more break the current PRGM (NOT a regular way for a BASIC !)
//
//      ADDED : CINIT <bauds> -> init the alt. serialPort
//      ADDED : COUT   -> send a byte to alt. serialPort
//      ADDED : CLIST  -> list to alt. serialPort
//      ADDED : CPRINT -> print to init the alt. serialPort
//      ADDED : CINV   -> invert UART Vs SoftwareSerial (but no for CINIT)
//
//      ADDED : CINP   -> waits a character from SLAVE serialPort
//
//
//
// v0.14: 2013-11-07
//      Input command always set the variable to 99
//      Modified Input command to accept an expression using getn()
//      Syntax is "input x" where x is any variable
//
// v0.13: 2013-03-04
//      Support for Arduino 1.5 (SPI.h included, additional changes for DUE support)
//
// v0.12: 2013-03-01
//      EEPROM load and save routines added: EFORMAT, ELIST, ELOAD, ESAVE, ECHAIN
//      added EAUTORUN option (chains to EEProm saved program on startup)
//      Bugfixes to build properly on non-arduino systems (PROGMEM #define workaround)
//      cleaned up a bit of the #define options wrt TONE
//
// v0.11: 2013-02-20
//      all display strings and tables moved to PROGMEM to save space
//      removed second serial
//      removed pinMode completely, autoconf is explicit
//      beginnings of EEPROM related functionality (new,load,save,list)
//
// v0.10: 2012-10-15
//      added kAutoConf, which eliminates the "PINMODE" statement.
//      now, DWRITE,DREAD,AWRITE,AREAD automatically set the PINMODE appropriately themselves.
//      should save a few bytes in your programs.
//
// v0.09: 2012-10-12
//      Fixed directory listings.  FILES now always works. (bug in the SD library)
//      ref: http://arduino.cc/forum/index.php/topic,124739.0.html
//      fixed filesize printouts (added printUnum for unsigned numbers)
//      #defineable baud rate for slow connection throttling
//
// v0.08: 2012-10-02
//      Tone generation through piezo added (TONE, TONEW, NOTONE)
//
// v0.07: 2012-09-30
//      Autorun buildtime configuration feature
//
// v0.06: 2012-09-27
//      Added optional second serial input, used for an external keyboard
//
// v0.05: 2012-09-21
//      CHAIN to load and run a second file
//      RND,RSEED for random stuff
//      Added "!=" for "<>" synonym
//      Added "END" for "STOP" synonym (proper name for the functionality anyway)
//
// v0.04: 2012-09-20

CustomSoftwareSerial_xtase.cpp
/*
CustomSoftwareSerialXts.cpp (formerly NewSoftSerial.cpp) -
Multi-instance software serial library for Arduino/Wiring
-- Interrupt-driven receive and other improvements by ladyada
   (http://ladyada.net)
-- Tuning, circular buffer, derivation from class Print/Stream,
   multi-instance support, porting to 8MHz processors,
   various optimizations, PROGMEM delay tables, inverse logic and
   direct port writing by Mikal Hart (http://www.arduiniana.org)
-- Pin change interrupt macros by Paul Stoffregen (http://www.pjrc.com)
-- 20MHz processor support by Garrett Mace (http://www.macetech.com)
-- ATmega1280/2560 support by Brett Hagman (http://www.roguerobotics.com/)

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

The latest version of this library can always be found at
http://arduiniana.org.
*/

// When set, _DEBUG co-opts pins 11 and 13 for debugging with an
// oscilloscope or logic analyzer.  Beware: it also slightly modifies
// the bit times, so don't rely on it too much at high baud rates
#define _DEBUG 0
#define _DEBUG_PIN1 11
#define _DEBUG_PIN2 13
//
// Includes
//
#include <avr/interrupt.h>
#include <avr/pgmspace.h>
#include <Arduino.h>
#include "CustomSoftwareSerial_xtase.h"
#include <math.h>
//
// Lookup table
//
typedef struct _DELAY_TABLE
{
  long baud;
  unsigned short rx_delay_centering;
  unsigned short rx_delay_intrabit;
  unsigned short rx_delay_stopbit;
  unsigned short tx_delay;
} DELAY_TABLE;

#if F_CPU == 16000000

static const DELAY_TABLE PROGMEM table[] =
{
  //  baud    rxcenter   rxintra    rxstop    tx

// MOA fix try -- doesn't works better
//  { 115200,   1,         17,        17,       12,    },
  { 115200,   0,          8,        8,     6,     },
//  { 57600,    1,          16,        16,     14     },
  { 57600,    10,        37,        37,       33,    },
// ========================================================
  
  { 38400,    25,        57,        57,       54,    },
  { 31250,    31,        70,        70,       68,    },
  { 28800,    34,        77,        77,       74,    },
  { 19200,    54,        117,       117,      114,   },
  { 14400,    74,        156,       156,      153,   },
  { 9600,     114,       236,       236,      233,   },

// MOA try => 4800 values x0.6
{ 8000,     139,       284,       284,      281,   },
  
  { 4800,     233,       474,       474,      471,   },
  { 2400,     471,       950,       950,      947,   },
  { 1200,     947,       1902,      1902,     1899,  },
  { 600,      1902,      3804,      3804,     3800,  },
  { 300,      3804,      7617,      7617,     7614,  },
};

const int XMIT_START_ADJUSTMENT = 5;

#elif F_CPU == 8000000

static const DELAY_TABLE table[] PROGMEM =
{
  //  baud    rxcenter    rxintra    rxstop  tx
  { 115200,   1,          5,         5,      3,      },
  { 57600,    1,          15,        15,     13,     },
  { 38400,    2,          25,        26,     23,     },
  { 31250,    7,          32,        33,     29,     },
  { 28800,    11,         35,        35,     32,     },
  { 19200,    20,         55,        55,     52,     },
  { 14400,    30,         75,        75,     72,     },
  { 9600,     50,         114,       114,    112,    },
  { 4800,     110,        233,       233,    230,    },
  { 2400,     229,        472,       472,    469,    },
  { 1200,     467,        948,       948,    945,    },
  { 600,      948,        1895,      1895,   1890,   },
  { 300,      1895,       3805,      3805,   3802,   },
};

const int XMIT_START_ADJUSTMENT = 4;

#elif F_CPU == 20000000

// 20MHz support courtesy of the good people at macegr.com.
// Thanks, Garrett!

static const DELAY_TABLE PROGMEM table[] =
{
  //  baud    rxcenter    rxintra    rxstop  tx
  { 115200,   3,          21,        21,     18,     },
  { 57600,    20,         43,        43,     41,     },
  { 38400,    37,         73,        73,     70,     },
  { 31250,    45,         89,        89,     88,     },
  { 28800,    46,         98,        98,     95,     },
  { 19200,    71,         148,       148,    145,    },
  { 14400,    96,         197,       197,    194,    },
  { 9600,     146,        297,       297,    294,    },
  { 4800,     296,        595,       595,    592,    },
  { 2400,     592,        1189,      1189,   1186,   },
  { 1200,     1187,       2379,      2379,   2376,   },
  { 600,      2379,       4759,      4759,   4755,   },
  { 300,      4759,       9523,      9523,   9520,   },
};

const int XMIT_START_ADJUSTMENT = 6;

#else

#error This version of CustomSoftwareSerialXts supports only 20, 16 and 8MHz processors

#endif

//
// Statics
//
CustomSoftwareSerialXts *CustomSoftwareSerialXts::active_object = 0;
char CustomSoftwareSerialXts::_receive_buffer[_SS_MAX_RX_BUFF];
volatile uint8_t CustomSoftwareSerialXts::_receive_buffer_tail = 0;
volatile uint8_t CustomSoftwareSerialXts::_receive_buffer_head = 0;


//
// Debugging
//
// This function generates a brief pulse
// for debugging or measuring on an oscilloscope.
inline void DebugPulse(uint8_t pin, uint8_t count)
{
#if _DEBUG
  volatile uint8_t *pport = portOutputRegister(digitalPinToPort(pin));

  uint8_t val = *pport;
  while (count--)
  {
    *pport = val | digitalPinToBitMask(pin);
    *pport = val;
  }
#endif
}

//
// Private methods
//

/* static */
inline void CustomSoftwareSerialXts::tunedDelay(uint16_t delay) {
  uint8_t tmp=0;

  asm volatile("sbiw    %0, 0x01 \n\t"
    "ldi %1, 0xFF \n\t"
    "cpi %A0, 0xFF \n\t"
    "cpc %B0, %1 \n\t"
    "brne .-10 \n\t"
    : "+r" (delay), "+a" (tmp)
    : "0" (delay)
    );
}

// This function sets the current object as the "listening"
// one and returns true if it replaces another
bool CustomSoftwareSerialXts::listen()
{
  if (active_object != this)
  {
    _buffer_overflow = false;
    uint8_t oldSREG = SREG;
    cli();
    _receive_buffer_head = _receive_buffer_tail = 0;
    active_object = this;
    SREG = oldSREG;
    return true;
  }

  return false;
}

//
// The receive routine called by the interrupt handler
//
void CustomSoftwareSerialXts::recv()
{

#if GCC_VERSION < 40302
// Work-around for avr-gcc 4.3.0 OSX version bug
// Preserve the registers that the compiler misses
// (courtesy of Arduino forum user *etracer*)
  asm volatile(
    "push r18 \n\t"
    "push r19 \n\t"
    "push r20 \n\t"
    "push r21 \n\t"
    "push r22 \n\t"
    "push r23 \n\t"
    "push r26 \n\t"
    "push r27 \n\t"
    ::);
#endif

  uint8_t d = 0;

  // If RX line is high, then we don't see any start bit
  // so interrupt is probably not for us
  if (_inverse_logic ? rx_pin_read() : !rx_pin_read())
  {
    // Wait approximately 1/2 of a bit width to "center" the sample
    tunedDelay(_rx_delay_centering);
    DebugPulse(_DEBUG_PIN2, 1);

    for (uint8_t i=0x1; i && i <= this->_maxValueOfDataBit; i <<= 1)
    {
      tunedDelay(_rx_delay_intrabit);
      DebugPulse(_DEBUG_PIN2, 1);
      uint8_t noti = ~i;
      if (rx_pin_read()) {
        d |= i;
      } else { // else clause added to ensure function timing is ~balanced
        d &= noti;
      }
    }

    // skip the parity bit
    if(this->_parityBit != NONE) {
        tunedDelay(_rx_delay_stopbit);
        DebugPulse(_DEBUG_PIN2, 1);
    }

    // skip the stop bit
    for(uint8_t i = 0; i < this->_numberOfStopBit; i ++) {
        tunedDelay(_rx_delay_stopbit);
        DebugPulse(_DEBUG_PIN2, 1);
    }

    if (_inverse_logic)
      d = ~d;

    // if buffer full, set the overflow flag and return
    if ((_receive_buffer_tail + 1) % _SS_MAX_RX_BUFF != _receive_buffer_head)
    {
      // save new data in buffer: tail points to where byte goes
      _receive_buffer[_receive_buffer_tail] = d; // save new byte
      _receive_buffer_tail = (_receive_buffer_tail + 1) % _SS_MAX_RX_BUFF;
    }
    else
    {
#if _DEBUG // for scope: pulse pin as overflow indictator
      DebugPulse(_DEBUG_PIN1, 1);
#endif
      _buffer_overflow = true;
    }
  }

#if GCC_VERSION < 40302
// Work-around for avr-gcc 4.3.0 OSX version bug
// Restore the registers that the compiler misses
  asm volatile(
    "pop r27 \n\t"
    "pop r26 \n\t"
    "pop r23 \n\t"
    "pop r22 \n\t"
    "pop r21 \n\t"
    "pop r20 \n\t"
    "pop r19 \n\t"
    "pop r18 \n\t"
    ::);
#endif
}

void CustomSoftwareSerialXts::tx_pin_write(uint8_t pin_state)
{
  if (pin_state == LOW)
    *_transmitPortRegister &= ~_transmitBitMask;
  else
    *_transmitPortRegister |= _transmitBitMask;
}

uint8_t CustomSoftwareSerialXts::rx_pin_read()
{
  return *_receivePortRegister & _receiveBitMask;
}

//
// Interrupt handling
//

/* static */
inline void CustomSoftwareSerialXts::handle_interrupt()
{
  if (active_object)
  {
    active_object->recv();
  }
}

#if defined(PCINT0_vect)
ISR(PCINT0_vect)
{
  CustomSoftwareSerialXts::handle_interrupt();
}
#endif

#if defined(PCINT1_vect)
ISR(PCINT1_vect)
{
  CustomSoftwareSerialXts::handle_interrupt();
}
#endif

#if defined(PCINT2_vect)
ISR(PCINT2_vect)
{
  CustomSoftwareSerialXts::handle_interrupt();
}
#endif

#if defined(PCINT3_vect)
ISR(PCINT3_vect)
{
  CustomSoftwareSerialXts::handle_interrupt();
}
#endif

//
// Constructor
//
CustomSoftwareSerialXts::CustomSoftwareSerialXts(uint8_t receivePin, uint8_t transmitPin, bool inverse_logic /* = false */) :
  _rx_delay_centering(0),
  _rx_delay_intrabit(0),
  _rx_delay_stopbit(0),
  _tx_delay(0),
  _buffer_overflow(false),
  _inverse_logic(inverse_logic)
{
  setTX(transmitPin);
  setRX(receivePin);
  setPort(CSERIAL_8N1);
}

//
// Destructor
//
CustomSoftwareSerialXts::~CustomSoftwareSerialXts()
{
  end();
}

void CustomSoftwareSerialXts::setTX(uint8_t tx)
{
  pinMode(tx, OUTPUT);
  digitalWrite(tx, _inverse_logic ? LOW : HIGH);
  _transmitBitMask = digitalPinToBitMask(tx);
  uint8_t port = digitalPinToPort(tx);
  _transmitPortRegister = portOutputRegister(port);
}

void CustomSoftwareSerialXts::setRX(uint8_t rx)
{
  pinMode(rx, INPUT);
  if (!_inverse_logic)
    digitalWrite(rx, HIGH);  // pullup for normal logic!
  _receivePin = rx;
  _receiveBitMask = digitalPinToBitMask(rx);
  uint8_t port = digitalPinToPort(rx);
  _receivePortRegister = portInputRegister(port);
}

//
// Public methods
//


// Xts Xts Xts Xts Xts Xts Xts Xts 
  // BEWARE : must be called before any Ops
  void CustomSoftwareSerialXts::setInvertLogic( bool inverse_logic ) {
    _inverse_logic = inverse_logic;
  }
// Xts Xts Xts Xts Xts Xts Xts Xts 

void CustomSoftwareSerialXts::begin(long speed)
{
  _rx_delay_centering = _rx_delay_intrabit = _rx_delay_stopbit = _tx_delay = 0;

  for (unsigned i=0; i<sizeof(table)/sizeof(table[0]); ++i)
  {
    long baud = pgm_read_dword(&table[i].baud);
    if (baud == speed)
    {
      _rx_delay_centering = pgm_read_word(&table[i].rx_delay_centering);
      _rx_delay_intrabit = pgm_read_word(&table[i].rx_delay_intrabit);
      _rx_delay_stopbit = pgm_read_word(&table[i].rx_delay_stopbit);
      _tx_delay = pgm_read_word(&table[i].tx_delay);
      break;
    }
  }

  // Set up RX interrupts, but only if we have a valid RX baud rate
  if (_rx_delay_stopbit)
  {
    if (digitalPinToPCICR(_receivePin))
    {
      *digitalPinToPCICR(_receivePin) |= _BV(digitalPinToPCICRbit(_receivePin));
      *digitalPinToPCMSK(_receivePin) |= _BV(digitalPinToPCMSKbit(_receivePin));
    }
    tunedDelay(_tx_delay); // if we were low this establishes the end
  }

#if _DEBUG
  pinMode(_DEBUG_PIN1, OUTPUT);
  pinMode(_DEBUG_PIN2, OUTPUT);
#endif

  listen();
}

void CustomSoftwareSerialXts::begin(long speed, uint16_t configuration) {
    this->setPort(configuration);
    this->begin(speed);
}

void CustomSoftwareSerialXts::end()
{
  if (digitalPinToPCMSK(_receivePin))
    *digitalPinToPCMSK(_receivePin) &= ~_BV(digitalPinToPCMSKbit(_receivePin));
}


// Read data from buffer
int CustomSoftwareSerialXts::read()
{
  if (!isListening())
    return -1;

  // Empty buffer?
  if (_receive_buffer_head == _receive_buffer_tail)
    return -1;

  // Read from "head"
  uint8_t d = _receive_buffer[_receive_buffer_head]; // grab next byte
  _receive_buffer_head = (_receive_buffer_head + 1) % _SS_MAX_RX_BUFF;
  return d;
}

int CustomSoftwareSerialXts::available()
{
  if (!isListening())
    return 0;

  return (_receive_buffer_tail + _SS_MAX_RX_BUFF - _receive_buffer_head) % _SS_MAX_RX_BUFF;
}

size_t CustomSoftwareSerialXts::write(uint8_t b)
{
  if (_tx_delay == 0) {
    setWriteError();
    return 0;
  }

  uint8_t oldSREG = SREG;
  cli();  // turn off interrupts for a clean txmit

  // Write the start bit
  tx_pin_write(_inverse_logic ? HIGH : LOW);
  tunedDelay(_tx_delay + XMIT_START_ADJUSTMENT);

  uint8_t numberOfBits1 = calculateNumberOfBits1(b);
  uint8_t maxValueOfData = round(pow(2, this->_numberOfDataBit - 1));

  // Write each of the 8 bits
  if (_inverse_logic)
  {
    for (byte mask = 0x01; mask && mask <= this->_maxValueOfDataBit; mask <<= 1)
    {
      if (b & mask) // choose bit
        tx_pin_write(LOW); // send 1
      else
        tx_pin_write(HIGH); // send 0

      tunedDelay(_tx_delay);
    }

  }
  else
  {
    for (byte mask = 0x01; mask && mask <= this->_maxValueOfDataBit; mask <<= 1)
    {
      if (b & mask) // choose bit
        tx_pin_write(HIGH); // send 1
      else
        tx_pin_write(LOW); // send 0

      tunedDelay(_tx_delay);
    }

  }

  writeParityBits(numberOfBits1);
// MOA
  // tunedDelay(_tx_delay);
  if(this->_parityBit != NONE) {
    tunedDelay(_tx_delay);
  }
// MOA
  writeStopBits();

  SREG = oldSREG; // turn interrupts back on
  tunedDelay(_tx_delay);

  return 1;
}

void CustomSoftwareSerialXts::flush()
{
  if (!isListening())
    return;

  uint8_t oldSREG = SREG;
  cli();
  _receive_buffer_head = _receive_buffer_tail = 0;
  SREG = oldSREG;
}

int CustomSoftwareSerialXts::peek()
{
  if (!isListening())
    return -1;

  // Empty buffer?
  if (_receive_buffer_head == _receive_buffer_tail)
    return -1;

  // Read from "head"
  return _receive_buffer[_receive_buffer_head];
}

void CustomSoftwareSerialXts::setPort(uint16_t configuration) {
    this->_numberOfDataBit = (uint8_t)(configuration / 100);
    this->_maxValueOfDataBit = round(pow(2, this->_numberOfDataBit - 1));
    this->_parityBit =       (Parity)(configuration % 100 / 10);
    this->_numberOfStopBit = (uint8_t)(configuration % 10);
}

uint8_t CustomSoftwareSerialXts::getNumberOfDataBit() {
    return this->_numberOfDataBit;
}

Parity CustomSoftwareSerialXts::getParity() {
    return this->_parityBit;
}

uint8_t CustomSoftwareSerialXts::getNumberOfStopBit() {
    return this->_numberOfStopBit;
}

void CustomSoftwareSerialXts::writeParityBits(uint8_t numberOfBit1) {
    if(this->_parityBit == NONE) {
        return;
    }

    if((this->_parityBit == EVEN &&  (numberOfBit1 & 0x01) && !_inverse_logic) ||
       (this->_parityBit == EVEN && !(numberOfBit1 & 0x01) &&  _inverse_logic) ||
       (this->_parityBit == ODD  &&  (numberOfBit1 & 0x01) &&  _inverse_logic) ||
       (this->_parityBit == ODD  && !(numberOfBit1 & 0x01) && !_inverse_logic)) {
        tx_pin_write(HIGH); // send 1
    }

    if((this->_parityBit == ODD  &&  (numberOfBit1 & 0x01) && !_inverse_logic) ||
       (this->_parityBit == ODD  && !(numberOfBit1 & 0x01) &&  _inverse_logic) ||
       (this->_parityBit == EVEN &&  (numberOfBit1 & 0x01) &&  _inverse_logic) ||
       (this->_parityBit == EVEN && !(numberOfBit1 & 0x01) && !_inverse_logic)) {
        tx_pin_write(LOW); // send 1
    }



    if(this->_parityBit == ODD) {
        if(numberOfBit1 & 0x01 && !_inverse_logic) {
          tx_pin_write(LOW); // send 0
          return;
        }

        if(numberOfBit1 & 0x01 && _inverse_logic) {
          tx_pin_write(HIGH); // send 1
          return;
        }

    }
}

void CustomSoftwareSerialXts::writeStopBits() {
    for(int8_t stopBitIndex = 0; stopBitIndex < this->_numberOfStopBit; stopBitIndex++) {
        if(_inverse_logic) {
            tx_pin_write(LOW);
        } else {
            tx_pin_write(HIGH);
        }

// MOA
tunedDelay(_tx_delay);
// MOA  
    }
}

uint8_t CustomSoftwareSerialXts::calculateNumberOfBits1(uint8_t sentData) {
    uint8_t numberOfBit1 = 0;
    uint8_t index;
    for (index = 0x80; index; index >>= 1) {
        if (sentData & index) numberOfBit1++;
    }

    return numberOfBit1;
}
